packages/cardano-services/src/Asset/types.ts

```diff
@@ -16,3 +17,15 @@ export interface NftMetadataService {
    */
   getNftMetadata(asset: AssetPolicyIdAndName): Promise<Asset.NftMetadata | undefined>;
 }
+
+/**
+ * Service to get CIP-38? token metadata for a given subject
+ */
+export interface TokenMetadataService extends Shutdown {
+  /**
+   * Get CIP-38? token metadata for a given subject
+   *
+   * @returns CIP-38? token metadata, `null` if not found
+   */
+  getTokenMetadata(assetIds: Cardano.AssetId[]): Promise<(Asset.TokenMetadata | null | undefined)[]>;
```

Comment:
Please document what `undefined` means. Seems to be indicating an error. Wouldn't throwing an error be more suitable instead of returning undefineds then?

Reply:
Yes @mkazlauskas ,

I taken [this line](https://github.com/input-output-hk/cardano-js-sdk/blob/75acb80b2310a183dca71b94d091b422d5ca6672/packages/core/src/Asset/types/AssetInfo.ts#L26) as specification: the only reason I can see why a token metadata could be not loaded is in case of error; so in case of error I'm returning `undefined` and logging the error.

But sorry, in your comment you are asking to better document the reason for `undefined` and proposing to change the interface to throw an error (which would also mean to remove `undefined` from possible returned values): the two actions are in contrast; which one is the preferred?

Thank you

Reply:
The line is only relevant on client-side. Token metadata was loaded async, so `undefined` could be used to distinguish assets that are:
- Not tokens (no token metadata) = `null`
- Don't know if it's a token or not (still loading) = `undefined`

Reply:
Sorry @mkazlauskas ,

applying this change I need to change [this line](https://github.com/input-output-hk/cardano-js-sdk/blob/feat/ADP-1605-token-metadata-provider/packages/cardano-services/src/Asset/CardanoTokenRegistry.ts#L123) with a `throw` statement: could you please suggest me the right error to throw?

Thank you

Reply:
Probably ProviderError{Unknown} that has real error in 'innerError' if it's just some internal error

Reply:
More or less the same question for the errors in the two `catch` blocks, @mkazlauskas ,

if I remember correctly we usually do not simply propagate the errors, but we encapsulate them in a SDK error type (at least once I used `ProviderError`). Should I use the same pattern?

Reply:
Resolved in cc5145c3eaf007707b35a34896ff093b42a30a85

could you please review this again @mkazlauskas  ? Thank you